File: golang/contains_duplicate_ii.go
package main

// Space : O(n)
// Time  : O(n)

func containsNearbyDuplicate(nums []int, k int) bool {
    n := len(nums)
    d := make(map[int]int)
    
    for i := 0; i < n; i++ {
        if _, ok := d[nums[i]]; !ok {
            d[nums[i]] = i
        } else {
            if i - d[nums[i]] <= k {
                return true
            }
            d[nums[i]] = i
        }
    } 
    
    return false
}

File: python3/contains_duplicate_ii.py
# Space : O(n)
# Time  : O(n)

class Solution:
    def containsNearbyDuplicate(self, nums: List[int], k: int) -> bool:
        d = {}

        for i, v in enumerate(nums):
            if v not in d:
                d[v] = i
            else:
                if i - d[v] <= k:
                    return True
                d[v] = i

        return False
